File: lib/agents/odds-selector.ts
import { renderMission } from '@/lib/agent-kernel'
import type { Blackboard } from '@/lib/agent-kernel/blackboard'
import type { AgentMission } from '@/lib/agent-kernel/types'
import type { PickCandidate, Tier, ReliablePick, ExcludedPick, TierCombo, TierDecision, OddsSelectorOutput } from '@/lib/types'
import { getBookmakerQuotes, PRIORITY_BOOKMAKER_KEY } from '@/lib/tools/odds-api'

// Appelé en interne par lib/agents/analyst.ts:runAnalystAndOdds — un seul
// appel visible depuis l'orchestrateur pour Analyste + Sélecteur de cotes,
// mais ce module reste séparé et 100% déterministe (aucun appel modèle).
const MISSION: AgentMission = {
  role: 'odds-selector',
  label: 'le Sélecteur de cotes',
  responsibility:
    "décider, à partir des picks candidats de l'Analyste, lesquels ont une cote fiable (consensus bookmakers) et comment ils composent les 3 combinés finaux (prudent/équilibré/audacieux) — décision finale sur ce qui est réellement publié.",
  doesNot: [
    "Ne sélectionne aucun nouveau match et ne calcule aucune tendance statistique — délégué à l'Analyste.",
    'Ne rédige pas le post Telegram — délégué au Rédacteur.',
    "Ne valide pas la qualité finale avant publication — cette décision revient à l'utilisateur depuis le dashboard.",
  ],
}

const MAX_BOOKMAKER_SPREAD = 0.20

// Un combiné perd dès qu'UN pick perd — le NOMBRE de picks contrôle donc la
// probabilité de gain final bien plus que la cote individuelle. Avec des
// picks à ~85% de confiance chacun, un combiné de 15 picks ne gagne que
// ~9% du temps (0.85^15) même si chaque pick est excellent individuellement
// — incompatible avec un palier "prudent" censé gagner majoritairement.
// Les 3 paliers sont donc différenciés par DEUX critères combinés :
// - le nombre de picks (ce qui pilote la probabilité de gain du combiné)
// - la cote individuelle de chaque pick (prudent = basse, audacieux = haute)
// prudent  : peu de picks, cote basse par pick  → gagne la majorité du temps
// équilibré: nombre intermédiaire, cotes mixtes → compromis
// audacieux: plus de picks et/ou cotes hautes   → rare, mais gros gain
export const TIER_PICK_RANGE: Record<Tier, { min: number; max: number }> = {
  prudent: { min: 2, max: 4 },
  equilibre: { min: 5, max: 8 },
  audacieux: { min: 8, max: 15 },
}
// Minimum absolu pour tenter de construire ne serait-ce qu'un seul palier —
// le plus bas des trois seuils ci-dessus (prudent).
export const MIN_PICKS_PER_COMBO = TIER_PICK_RANGE.prudent.min

function median(values: number[]): number {
  const sorted = [...values].sort((a, b) => a - b)
  const mid = Math.floor(sorted.length / 2)
  return sorted.length % 2 === 0 ? (sorted[mid - 1] + sorted[mid]) / 2 : sorted[mid]
}

function matchLabel(pick: PickCandidate): string {
  return `${pick.home_team} - ${pick.away_team} (${pick.bet_type})`
}

// Identifie le MATCH (pas le pick) — l'Analyste peut proposer plusieurs
// prédictions indépendantes sur un même match (ex: "Over 2.5" et "BTTS
// Oui" pour PSG-Barça), mais un seul de ces picks doit finir dans UN
// combiné donné : deux paris sur le même match dans le même coupon sont
// des résultats corrélés, pas une vraie diversification.
function matchKey(pick: PickCandidate): string {
  return `${pick.home_team}-${pick.away_team}`
}

/**
 * ÉTAPE 1 — cote fiable par pick candidat. Un pick déjà validé par
 * l'Analyste peut être rejeté ici si le marché est jugé trop incertain
 * (écart entre bookmakers > 20%) — le Sélecteur de cotes a le dernier mot.
 */
async function selectReliableOdds(
  candidates: PickCandidate[]
): Promise<{ reliable: ReliablePick[]; excluded: ExcludedPick[] }> {
  const reliable: ReliablePick[] = []
  const excluded: ExcludedPick[] = []

  for (const pick of candidates) {
    const quotes = await getBookmakerQuotes(pick.home_team, pick.away_team, pick.bet_type)

    if (!quotes.length) {
      excluded.push({ match: matchLabel(pick), bet_type: pick.bet_type, reason: 'Aucune cote bookmaker disponible pour ce marché' })
      continue
    }

    const prices = quotes.map(q => q.price)
    const min = Math.min(...prices)
    const max = Math.max(...prices)
    const spread = (max - min) / min

    if (spread > MAX_BOOKMAKER_SPREAD) {
      excluded.push({
        match: matchLabel(pick),
        bet_type: pick.bet_type,
        reason: `Écart entre bookmakers ${(spread * 100).toFixed(1)}% > ${MAX_BOOKMAKER_SPREAD * 100}% — marché trop incertain`,
      })
      continue
    }

    const priority = quotes.find(q => q.bookmaker === PRIORITY_BOOKMAKER_KEY)
    const finalOdds = priority ? priority.price : median(prices)
    const source = priority ? '1xBet (priorité 1)' : `médiane sur ${quotes.length} bookmakers`

    reliable.push({ ...pick, odds: finalOdds, odds_source: source, bookmaker_spread_pct: Math.round(spread * 1000) / 10 })
  }

  return { reliable, excluded }
}

function combinedOddsOf(picks: ReliablePick[]): number {
  return Math.round(picks.reduce((acc, p) => acc * p.odds, 1) * 100) / 100
}

// Un seul pick par match dans un même combiné — voir matchKey().
function takeUniqueByMatch(sorted: ReliablePick[], count: number): ReliablePick[] {
  const result: ReliablePick[] = []
  const seen = new Set<string>()
  for (const pick of sorted) {
    if (result.length >= count) break
    const key = matchKey(pick)
    if (seen.has(key)) continue
    seen.add(key)
    result.push(pick)
  }
  return result
}

/** Alterne du bas puis du haut d'une liste triée par cote croissante — mélange équilibré des risques. */
function interleaveFromBothEnds(sortedAsc: ReliablePick[], count: number): ReliablePick[] {
  const result: ReliablePick[] = []
  const seen = new Set<string>()
  let lo = 0
  let hi = sortedAsc.length - 1
  let takeLow = true

  while (result.length < count && lo <= hi) {
    const candidate = takeLow ? sortedAsc[lo] : sortedAsc[hi]
    if (takeLow) lo++
    else hi--

    const key = matchKey(candidate)
    if (seen.has(key)) continue // retente le même côté, pointeur déjà avancé

    seen.add(key)
    result.push(candidate)
    takeLow = !takeLow
  }

  return result
}

function buildCombo(tier: Tier, picks: ReliablePick[], reasonSuffix: string, decisions: TierDecision[]): TierCombo {
  for (const p of picks) {
    decisions.push({ match: matchLabel(p), tier, included: true, reason: `Sélectionné — ${reasonSuffix}` })
  }
  return { tier, picks, combined_odds: combinedOddsOf(picks) }
}

/**
 * ÉTAPE 2 — composition des 3 combinés. Chaque palier a sa propre plage de
 * nombre de picks (TIER_PICK_RANGE) — pas la même pour les 3 — en plus de sa
 * propre logique de cote : prudent pioche les cotes individuelles les plus
 * basses (et le moins de picks, pour maximiser la probabilité de gain du
 * combiné), audacieux les plus hautes (et le plus de picks, accepte un gain
 * rare mais élevé), équilibré un mélange des deux à un volume intermédiaire.
 */
function composeTiers(reliable: ReliablePick[]): { combos: Partial<Record<Tier, TierCombo>>; decisions: TierDecision[] } {
  const combos: Partial<Record<Tier, TierCombo>> = {}
  const decisions: TierDecision[] = []

  const byOddsAsc = [...reliable].sort((a, b) => a.odds - b.odds)
  const byOddsDesc = [...byOddsAsc].reverse()
  // Le vrai plafond, c'est le nombre de MATCHS distincts — plusieurs picks
  // fiables sur le même match (ex: Over 2.5 ET BTTS Oui pour PSG-Barça) ne
  // comptent qu'une fois, un seul finira dans un combiné donné.
  const uniqueMatchCount = new Set(reliable.map(matchKey)).size

  for (const tier of ['prudent', 'equilibre', 'audacieux'] as Tier[]) {
    const { min, max } = TIER_PICK_RANGE[tier]

    if (uniqueMatchCount < min) {
      decisions.push({
        match: '—',
        tier,
        included: false,
        reason: `Seulement ${uniqueMatchCount} matchs distincts avec cote fiable — minimum ${min} requis pour le palier ${tier}`,
      })
      continue
    }

    const count = Math.min(max, uniqueMatchCount)

    if (tier === 'prudent') {
      combos.prudent = buildCombo('prudent', takeUniqueByMatch(byOddsAsc, count), `${count} picks à cote individuelle basse — vise à gagner majoritairement`, decisions)
    } else if (tier === 'audacieux') {
      combos.audacieux = buildCombo('audacieux', takeUniqueByMatch(byOddsDesc, count), `${count} picks à cote individuelle haute — gain rare mais élevé`, decisions)
    } else {
      combos.equilibre = buildCombo('equilibre', interleaveFromBothEnds(byOddsAsc, count), `${count} picks — mélange de cotes basses et hautes`, decisions)
    }
  }

  return { combos, decisions }
}

export async function decide(candidates: PickCandidate[], blackboard: Blackboard): Promise<OddsSelectorOutput> {
  const mission = renderMission(MISSION) // documente la mission dans le journal des agents, pas un prompt LLM — cet agent est déterministe
  blackboard.write('odds-selector-mission', mission)

  const { reliable, excluded } = await selectReliableOdds(candidates)
  blackboard.post({
    from: 'odds-selector',
    type: 'observation',
    content: `${reliable.length}/${candidates.length} picks avec cote fiable — ${excluded.length} rejetés (marché incertain ou cote absente).`,
  })

  const { combos, decisions } = composeTiers(reliable)
  const builtTiers = (Object.keys(combos) as Tier[]).filter(t => combos[t])

  for (const tier of ['prudent', 'equilibre', 'audacieux'] as Tier[]) {
    const combo = combos[tier]
    blackboard.post({
      from: 'odds-selector',
      to: 'writer',
      type: 'decision',
      content: combo
        ? `Palier ${tier} : ${combo.picks.length} picks, cote combinée ${combo.combined_odds}.`
        : `Palier ${tier} non généré aujourd'hui — moins de ${TIER_PICK_RANGE[tier].min} picks fiables disponibles.`,
    })
  }

  if (!builtTiers.length) {
    blackboard.post({ from: 'odds-selector', type: 'result', content: 'Aucun palier constructible aujourd\'hui.' })
  }

  return { reliable_picks: reliable, excluded_picks: excluded, combos, decisions }
}

File: backend/agents/odds_selector.py
"""Port de lib/agents/odds-selector.ts — 100% déterministe, aucun appel modèle.

Appelé en interne par agents/analyst.py:run_analyst_and_odds — un seul appel
visible depuis l'orchestrateur pour Analyste + Sélecteur de cotes.
"""
from agent_kernel import Blackboard, render_mission
from agent_kernel.types import AgentMission
from tools.odds_api import get_bookmaker_quotes, PRIORITY_BOOKMAKER_KEY

MISSION = AgentMission(
    role="odds-selector",
    label="le Sélecteur de cotes",
    responsibility=(
        "décider, à partir des picks candidats de l'Analyste, lesquels ont une cote fiable (consensus "
        "bookmakers) et comment ils composent les 3 combinés finaux (prudent/équilibré/audacieux) — "
        "décision finale sur ce qui est réellement publié."
    ),
    does_not=[
        "Ne sélectionne aucun nouveau match et ne calcule aucune tendance statistique — délégué à l'Analyste.",
        "Ne rédige pas le post Telegram — délégué au Rédacteur.",
        "Ne valide pas la qualité finale avant publication — cette décision revient à l'utilisateur depuis le dashboard.",
    ],
)

MAX_BOOKMAKER_SPREAD = 0.20

# Un combiné perd dès qu'UN pick perd — le NOMBRE de picks contrôle donc la
# probabilité de gain final bien plus que la cote individuelle.
# prudent  : peu de picks, cote basse par pick  → gagne la majorité du temps
# équilibré: nombre intermédiaire, cotes mixtes → compromis
# audacieux: plus de picks et/ou cotes hautes   → rare, mais gros gain
TIER_PICK_RANGE = {
    "prudent": {"min": 2, "max": 4},
    "equilibre": {"min": 5, "max": 8},
    "audacieux": {"min": 8, "max": 15},
}
MIN_PICKS_PER_COMBO = TIER_PICK_RANGE["prudent"]["min"]

ALL_TIERS = ["prudent", "equilibre", "audacieux"]


def _median(values: list[float]) -> float:
    s = sorted(values)
    n = len(s)
    mid = n // 2
    return (s[mid - 1] + s[mid]) / 2 if n % 2 == 0 else s[mid]


def _match_label(pick: dict) -> str:
    return f"{pick['home_team']} - {pick['away_team']} ({pick['bet_type']})"


def _match_key(pick: dict) -> str:
    """Identifie le MATCH (pas le pick) — l'Analyste peut proposer plusieurs
    prédictions indépendantes sur un même match, mais un seul de ces picks
    doit finir dans UN combiné donné."""
    return f"{pick['home_team']}-{pick['away_team']}"


async def _select_reliable_odds(candidates: list[dict]) -> dict:
    """ÉTAPE 1 — cote fiable par pick candidat. Un pick déjà validé par
    l'Analyste peut être rejeté ici si le marché est jugé trop incertain
    (écart entre bookmakers > 20%) — le Sélecteur de cotes a le dernier mot."""
    reliable: list[dict] = []
    excluded: list[dict] = []

    for pick in candidates:
        quotes = await get_bookmaker_quotes(pick["home_team"], pick["away_team"], pick["bet_type"])

        if not quotes:
            excluded.append({"match": _match_label(pick), "bet_type": pick["bet_type"], "reason": "Aucune cote bookmaker disponible pour ce marché"})
            continue

        prices = [q.price for q in quotes]
        lo, hi = min(prices), max(prices)
        spread = (hi - lo) / lo

        if spread > MAX_BOOKMAKER_SPREAD:
            excluded.append(
                {
                    "match": _match_label(pick),
                    "bet_type": pick["bet_type"],
                    "reason": f"Écart entre bookmakers {spread * 100:.1f}% > {MAX_BOOKMAKER_SPREAD * 100}% — marché trop incertain",
                }
            )
            continue

        priority = next((q for q in quotes if q.bookmaker == PRIORITY_BOOKMAKER_KEY), None)
        final_odds = priority.price if priority else _median(prices)
        source = "1xBet (priorité 1)" if priority else f"médiane sur {len(quotes)} bookmakers"

        reliable.append({**pick, "odds": final_odds, "odds_source": source, "bookmaker_spread_pct": round(spread * 1000) / 10})

    return {"reliable": reliable, "excluded": excluded}


def _combined_odds_of(picks: list[dict]) -> float:
    total = 1.0
    for p in picks:
        total *= p["odds"]
    return round(total * 100) / 100


def _take_unique_by_match(sorted_picks: list[dict], count: int) -> list[dict]:
    """Un seul pick par match dans un même combiné."""
    result: list[dict] = []
    seen: set[str] = set()
    for pick in sorted_picks:
        if len(result) >= count:
            break
        key = _match_key(pick)
        if key in seen:
            continue
        seen.add(key)
        result.append(pick)
    return result


def _interleave_from_both_ends(sorted_asc: list[dict], count: int) -> list[dict]:
    """Alterne du bas puis du haut d'une liste triée par cote croissante —
    mélange équilibré des risques."""
    result: list[dict] = []
    seen: set[str] = set()
    lo, hi = 0, len(sorted_asc) - 1
    take_low = True

    while len(result) < count and lo <= hi:
        candidate = sorted_asc[lo] if take_low else sorted_asc[hi]
        if take_low:
            lo += 1
        else:
            hi -= 1

        key = _match_key(candidate)
        if key in seen:
            continue  # retente le même côté, pointeur déjà avancé

        seen.add(key)
        result.append(candidate)
        take_low = not take_low

    return result


def _build_combo(tier: str, picks: list[dict], reason_suffix: str, decisions: list[dict]) -> dict:
    for p in picks:
        decisions.append({"match": _match_label(p), "tier": tier, "included": True, "reason": f"Sélectionné — {reason_suffix}"})
    return {"tier": tier, "picks": picks, "combined_odds": _combined_odds_of(picks)}


def _compose_tiers(reliable: list[dict]) -> dict:
    """ÉTAPE 2 — composition des 3 combinés. Chaque palier a sa propre plage
    de nombre de picks, en plus de sa propre logique de cote."""
    combos: dict[str, dict] = {}
    decisions: list[dict] = []

    by_odds_asc = sorted(reliable, key=lambda p: p["odds"])
    by_odds_desc = list(reversed(by_odds_asc))
    # Le vrai plafond, c'est le nombre de MATCHS distincts.
    unique_match_count = len({_match_key(p) for p in reliable})

    for tier in ALL_TIERS:
        rng = TIER_PICK_RANGE[tier]

        if unique_match_count < rng["min"]:
            decisions.append(
                {
                    "match": "—",
                    "tier": tier,
                    "included": False,
                    "reason": f"Seulement {unique_match_count} matchs distincts avec cote fiable — minimum {rng['min']} requis pour le palier {tier}",
                }
            )
            continue

        count = min(rng["max"], unique_match_count)

        if tier == "prudent":
            combos["prudent"] = _build_combo("prudent", _take_unique_by_match(by_odds_asc, count), f"{count} picks à cote individuelle basse — vise à gagner majoritairement", decisions)
        elif tier == "audacieux":
            combos["audacieux"] = _build_combo("audacieux", _take_unique_by_match(by_odds_desc, count), f"{count} picks à cote individuelle haute — gain rare mais élevé", decisions)
        else:
            combos["equilibre"] = _build_combo("equilibre", _interleave_from_both_ends(by_odds_asc, count), f"{count} picks — mélange de cotes basses et hautes", decisions)

    return {"combos": combos, "decisions": decisions}


async def decide(candidates: list[dict], blackboard: Blackboard) -> dict:
    mission = render_mission(MISSION)  # documente la mission dans le journal des agents, pas un prompt LLM
    blackboard.write("odds-selector-mission", mission)

    selection = await _select_reliable_odds(candidates)
    reliable, excluded = selection["reliable"], selection["excluded"]

    blackboard.post(
        from_role="odds-selector",
        type="observation",
        content=f"{len(reliable)}/{len(candidates)} picks avec cote fiable — {len(excluded)} rejetés (marché incertain ou cote absente).",
    )

    composition = _compose_tiers(reliable)
    combos, decisions = composition["combos"], composition["decisions"]
    built_tiers = [t for t in ALL_TIERS if t in combos]

    for tier in ALL_TIERS:
        combo = combos.get(tier)
        content = (
            f"Palier {tier} : {len(combo['picks'])} picks, cote combinée {combo['combined_odds']}."
            if combo
            else f"Palier {tier} non généré aujourd'hui — moins de {TIER_PICK_RANGE[tier]['min']} picks fiables disponibles."
        )
        blackboard.post(from_role="odds-selector", to_role="writer", type="decision", content=content)

    if not built_tiers:
        blackboard.post(from_role="odds-selector", type="result", content="Aucun palier constructible aujourd'hui.")

    return {"reliable_picks": reliable, "excluded_picks": excluded, "combos": combos, "decisions": decisions}
